kotlin-spark-api/2.4/src/main/kotlin/org/jetbrains/kotlinx/spark/api/UDFRegister.kt

```diff
@@ -0,0 +1,1356 @@
+@file:Suppress("DuplicatedCode")
+
+package org.jetbrains.kotlinx.spark.api
+
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.DataTypeWithClass
+import org.apache.spark.sql.UDFRegistration
+import org.apache.spark.sql.api.java.*
+import org.apache.spark.sql.functions
+import org.apache.spark.sql.types.DataType
+import scala.collection.mutable.WrappedArray
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.typeOf
+
+fun DataType.unWrapper(): DataType {
+    return when (this) {
+        is DataTypeWithClass -> DataType.fromJson(dt().json())
+        else -> this
+    }
+}
+
+/**
+ * Checks if [this] is of a valid type for an UDF, otherwise it throws a [TypeOfUDFParameterNotSupportedException]
+ */
+@PublishedApi
+internal fun KClass<*>.checkForValidType(parameterName: String) {
+    if (this == String::class || isSubclassOf(WrappedArray::class)) return // Most of the time we need strings or WrappedArrays
+    if (isSubclassOf(Iterable::class) || java.isArray
+        || isSubclassOf(Map::class) || isSubclassOf(Array::class)
+        || isSubclassOf(ByteArray::class) || isSubclassOf(CharArray::class)
+        || isSubclassOf(ShortArray::class) || isSubclassOf(IntArray::class)
+        || isSubclassOf(LongArray::class) || isSubclassOf(FloatArray::class)
+        || isSubclassOf(DoubleArray::class) || isSubclassOf(BooleanArray::class)
+    ) {
+        throw TypeOfUDFParameterNotSupportedException(this, parameterName)
+    }
+}
+
+/**
+ * An exception thrown when the UDF is generated with illegal types for the parameters
+ */
+class TypeOfUDFParameterNotSupportedException(kClass: KClass<*>, parameterName: String) : IllegalArgumentException(
+    "Parameter $parameterName is subclass of ${kClass.qualifiedName}. If you need to process an array use ${WrappedArray::class.qualifiedName}."
+)
+
+/**
+ * A wrapper for an UDF with 0 arguments.
+ * @property udfName the name of the UDF
+ */
+class UDFWrapper0(private val udfName: String) {
+    /**
+     * Calls the [functions.callUDF] for the UDF with the [udfName] and the given columns.
+     */
+    operator fun invoke(): Column {
+        return functions.callUDF(udfName)
+    }
+}
+
+/**
+ * Registers the [func] with its [name] in [this]
+ */
+@OptIn(ExperimentalStdlibApi::class)
+inline fun <reified R> UDFRegistration.register(name: String, noinline func: () -> R): UDFWrapper0 {
+    register(name, UDF0(func), schema(typeOf<R>()).unWrapper())
```

Comment:
It looks like it makes sense to work not with our generated schema, but with schema from an encoder, isn't it?

Reply:
Yes. Actually, it is better than hack `RowEncoder` to adapter our generated schema when register a UDF. reason: 1. `RowEncoder` only need a correctly `DataType`, not our generated schema other info. 2. some  spark action (eg: `.as<Int>()` action) to use UDF result is not though `RowEncoder`,if still use hack class way, we will hack many class and there is no guarantee that hacked classes will work well.